src/dtx/dtx_common.c

```diff
@@ -0,0 +1,528 @@
+		int	rc;
+
+		rc = vos_dtx_aggregate(cont->sc_hdl, DTX_AGG_YIELD_INTERVAL,
+				       DTX_AGG_THRESHOLD_AGE_LOWER);
+		if (rc != 0)
+			break;
+
+		ABT_thread_yield();
+	}
+
+	cont->sc_dtx_aggregating = 0;
+	ds_cont_put(cont);
+}
+
+static void
+dtx_flush_committable(struct ds_cont_hdl *cont_hdl)
+{
+	struct ds_cont		*cont = cont_hdl->sch_cont;
+	int			 rc;
+
+	do {
+		struct dtx_entry	*dte = NULL;
+
+		rc = vos_dtx_list_committable(cont->sc_hdl,
+					      DTX_THRESHOLD_COUNT, &dte);
+		if (rc <= 0)
+			break;
+
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+				dte, rc, cont_hdl->sch_pool->spc_map_version);
+		D_FREE(dte);
+	} while (rc >= 0);
+
+	if (rc < 0)
+		D_ERROR(DF_UUID": Fail to flush CoS cache: rc = %d\n",
+			DP_UUID(cont->sc_uuid), rc);
+
+	cont->sc_dtx_flushed = 1;
+}
+
+void
+dtx_batched_commit(void *arg)
+{
+	struct dss_module_info	*dmi = dss_get_module_info();
+
+	while (1) {
+		ABT_bool			 state;
+		struct dtx_batched_commit_args	*dbca;
+		struct ds_cont_hdl		*cont_hdl;
+		struct ds_cont			*cont;
+		struct dtx_entry		*dte = NULL;
+		struct dtx_stat			 stat = { 0 };
+		int				 rc;
+
+		if (d_list_empty(&dmi->dmi_dtx_batched_list))
+			goto check;
+
+		dbca = d_list_entry(dmi->dmi_dtx_batched_list.next,
+				    struct dtx_batched_commit_args, dbca_link);
+		cont_hdl = dbca->dbca_cont_hdl;
+		cont = cont_hdl->sch_cont;
+		if (cont->sc_closing) {
+			dtx_flush_committable(cont_hdl);
+			/* XXX: There may be resent-close during DTX flush,
+			 *	so NOT remove the @dbca from the list until
+			 *	DTX flushed.
+			 */
+			d_list_del(&dbca->dbca_link);
+			/* drop the ref for dtx_batched_commit_register(). */
+			ds_cont_hdl_put(cont_hdl);
+			D_FREE_PTR(dbca);
+			goto check;
+		}
+
+		d_list_move_tail(&dbca->dbca_link, &dmi->dmi_dtx_batched_list);
+		vos_dtx_stat(cont->sc_hdl, &stat);
+
+		if ((stat.dtx_committable_count > DTX_THRESHOLD_COUNT) ||
+		    (stat.dtx_oldest_committable_time != 0 &&
+		     dtx_hlc_age2sec(stat.dtx_oldest_committable_time) >
+		     DTX_COMMIT_THRESHOLD_AGE)) {
+			rc = vos_dtx_list_committable(cont->sc_hdl,
+						DTX_THRESHOLD_COUNT, &dte);
+			if (rc > 0) {
+				rc = dtx_commit(cont_hdl->sch_pool->spc_uuid,
+					cont->sc_uuid, dte, rc,
+					cont_hdl->sch_pool->spc_map_version);
+				D_FREE(dte);
+
+				if (cont->sc_closing) {
+					dtx_flush_committable(cont_hdl);
+					d_list_del(&dbca->dbca_link);
+					ds_cont_hdl_put(cont_hdl);
+					D_FREE_PTR(dbca);
+					goto check;
+				}
+
+				if (!cont->sc_dtx_aggregating)
+					vos_dtx_stat(cont->sc_hdl, &stat);
+			}
+		}
+
+		if (!cont->sc_dtx_aggregating &&
+		    ((stat.dtx_committed_count > DTX_AGG_THRESHOLD_CNT) ||
+		     (stat.dtx_oldest_committed_time != 0 &&
+		      dtx_hlc_age2sec(stat.dtx_oldest_committed_time) >
+				DTX_AGG_THRESHOLD_AGE_UPPER))) {
+			ds_cont_get(cont);
+			cont->sc_dtx_aggregating = 1;
+			rc = dss_ult_create(dtx_aggregate, cont, DSS_ULT_SELF,
+					    dmi->dmi_tgt_id, 0, NULL);
```

Comment:
dss_ult_create() changed a little on latest master, may need a rebasing. (may add a new ULT type or use DSS_ULT_MISC type - dss_ult_create(dtx_aggregate, cont, DSS_ULT_MISC, DSS_TGT_SELF, 0, NULL);

Reply:
I noticed your changes for that. I will rebase my patch.